### PR TITLE
Compiled log_softmax + gather operation to reduce vram requirements

### PR DESCRIPTION
**Pull request by tyler-romero**
The following operation is common in many post-training algorithms:
```python
all_logprobs = logits.log_softmax(dim=-1)  # shape: (batch_size, seq_len, vocab_size)
logprobs  = torch.gather(all_logprobs, dim=-1, index=labels.unsqueeze(-1)).squeeze(-1)
```

However it is fairly memory inefficient because `all_logprobs` is a large tensor (the same size as logits) that is only needed temporarily for the gather operation. For a modest batch size of 16, seq length of 1024, and vocab size of 32768 this tensor requires 2147 MB of vram (1073 MB for bfloat16).

torch.compile is able to optimize away this memory allocation so that a negligible amount of memory is required for this op other than to hold the logits in the first place.

if torch.compile() is not desired, there are other ways to achieve the same effect, such as:
```python
logsumexp_values = torch.stack([torch.logsumexp(l, dim=-1) for l in logits])
token_logits = torch.gather(logits, dim=-1, index=labels.unsqueeze(-1)).squeeze(-1)
logprobs = token_logits - logsumexp_values
```

### CODE REVIEW

**Comment by tyler-romero** [comment]
Benchmark:
```
import time
import torch

def naive_op(logits, index):
    logprobs = logits.log_softmax(dim=-1)  # shape: (batch_size, seq_len, vocab_size)
    return torch.gather(logprobs, dim=-1, index=index.unsqueeze(-1)).squeeze(-1)

@torch.compile(dynamic=True)
def compiled_op(logits, index):
    logprobs = logits.log_softmax(dim=-1)
    return torch.gather(logprobs, dim=-1, index=index.unsqueeze(-1)).squeeze(-1)

def measure_memory_and_time(func, logits, index, n_runs=100):
    torch.cuda.reset_peak_memory_stats()
    result = func(logits, index)
    mem_peak = torch.cuda.max_memory_allocated()
    start_time = time.perf_counter()
    for _ in range(n_runs):
        func(logits, index)
    avg_time = (time.perf_counter() - start_time) / n_runs
    return result, avg_time, mem_peak

# Simulated data
torch.manual_seed(42)
vocab_size = 32768
seq_len = 1024
batch_size = 16

device = "cuda" if torch.cuda.is_available() else "cpu"
logits = torch.randn(batch_size, seq_len, vocab_size, device=device, dtype=torch.float32)
index = torch.randint(0, vocab_size, (batch_size, seq_len), device=device)
logit_mem = torch.cuda.max_memory_allocated()

# Run all methods
naive_result, naive_time, naive_mem = measure_memory_and_time(naive_op, logits, index)
compiled_result, compiled_time, compiled_mem = measure_memory_and_time(compiled_op, logits, index)

# Check equivalence
print("Logits Dtype:", logits.dtype)
print("Memory consumed by logits: {:.2f} MB".format(logit_mem / 1e6))
print("Max absolute difference (naive and compiled):", (naive_result - compiled_result).abs().max().item())
print("Naive method time:      {:.6f} sec, Memory peak: {:.2f} MB".format(naive_time, naive_mem / 1e6))
print("Compiled method time:   {:.6f} sec, Memory peak: {:.2f} MB".format(compiled_time, compiled_mem / 1e6))
```

float32 results:
```
Logits Dtype: torch.float32
Memory consumed by logits: 2147.61 MB
Max absolute difference (naive and compiled): 9.5367431640625e-07
Naive method time:      0.000020 sec, Memory peak: 4295.16 MB
Compiled method time:   0.000066 sec, Memory peak: 2147.75 MB
```

bfloat16 results:
```
Logits Dtype: torch.bfloat16
Memory consumed by logits: 1073.87 MB
Max absolute difference (naive and compiled): 0.0
Naive method time:      0.000021 sec, Memory peak: 2147.65 MB
Compiled method time:   0.000068 sec, Memory peak: 1073.94 MB
```

**Review by vwxyzjn — APPROVED** [review]
LGTM. Thanks! Yeah I heard about how when doing fine-tune, you'd have to materialize the logits tensor and that's really expensive. Didn't know there is an easy API that just does this.

Merging as is. Thanks!

**Comment by gau-nernst** [comment]
I came across this function while looking at the open-instruct codebase. Isn't `log_softmax_and_gather` just negative cross entropy?

```python
from open_instruct.model_utils import log_softmax_and_gather
import torch
import torch.nn.functional as F

B, L, vocab_size = 4, 1024, 128_000

logits = torch.randn(B, L, vocab_size, device="cuda")
indices = torch.randint(128_000, size=(B, L), device="cuda")

out1 = log_softmax_and_gather(logits, indices)
out2 = -F.cross_entropy(logits.flatten(0, 1), indices.flatten(), reduction="none").view(B, L)

torch.testing.assert_close(out1, out2)
```
